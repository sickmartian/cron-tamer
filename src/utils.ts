import { CronSchedule, TimeSlot } from './types';
import { parseExpression } from 'cron-parser';

const COLORS = [
  '#FF6B6B', // red
  '#4ECDC4', // teal
  '#45B7D1', // blue
  '#96CEB4', // green
  '#FFEEAD', // yellow
  '#D4A5A5', // pink
  '#9B59B6', // purple
  '#3498DB', // blue
  '#E67E22', // orange
  '#2ECC71', // green
];

export function generateColor(index: number): string {
  return COLORS[index % COLORS.length];
}

export function parseCronExpression(
  expression: string,
  targetDate: Date = new Date(),
  timezone: string = 'UTC'
): Date[] {
  try {
    const paddedMonth = targetDate.getMonth() + 1 < 10 ? `0${targetDate.getMonth() + 1}` : targetDate.getMonth() + 1;
    const endOfMonth = new Date(targetDate.getFullYear(), targetDate.getMonth() + 1, 0);
    const paddedEndOfMonthDay = endOfMonth.getDate() < 10 ? `0${endOfMonth.getDate()}` : endOfMonth.getDate();
    const options = {
      currentDate: `${targetDate.getFullYear()}-${paddedMonth}-01T00:00:00`,
      tz: timezone,
      endDate: `${endOfMonth.getFullYear()}-${paddedMonth}-${paddedEndOfMonthDay}T23:59:59`,
    };
    
    const interval = parseExpression(expression, options);
    const occurrences: Date[] = [];
    
    try {
      let current = interval.next();
      while (true) {
        occurrences.push(current.toDate());
        current = interval.next();
      }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    } catch (err) {
      // This is expected when we reach the end of the month
      // The occurrences array will contain all valid dates up to that point
    }
    
    return occurrences;
  } catch (error) {
    console.error('Error parsing cron expression:', error);
    return [];
  }
}

export function generateTimeSlots(
  schedules: CronSchedule[],
  targetDate: Date,
  timezone: string = 'UTC'
): TimeSlot[] {
  const timeSlots: TimeSlot[] = [];

  schedules.forEach((schedule) => {
    if (!schedule.isActive) return;

    const occurrences = parseCronExpression(schedule.expression, targetDate, timezone);
    
    occurrences.forEach((start) => {
      timeSlots.push({
        start,
        duration: schedule.duration,
        scheduleId: schedule.id,
        scheduleName: schedule.name,
        key: `${schedule.id}-${start.getTime()}`
      });
    });
  });

  return timeSlots.sort((a, b) => a.start.getTime() - b.start.getTime());
}

export function getTimeSlotKey(scheduleId: string, startTime: Date): string {
  return `${scheduleId}-${startTime.getTime()}`;
} 