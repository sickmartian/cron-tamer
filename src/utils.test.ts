import { getOccurrecesRelevantForMonth, validateCronExpression } from "./utils";

describe("getOccurrecesRelevantForMonth", () => {
  it.each([
    // DST
    { case: 'DST', tz: "UTC", expression: "0 2 30 3 *" },
    { case: 'DST', tz: "UTC", expression: "30 2 30 3 *" },
    { case: 'DST', tz: "UTC", expression: "0 3 30 3 *" },
    { case: 'DST', tz: "UTC", expression: "30 3 30 3 *" },
    { case: 'DST', tz: "UTC", expression: "0 1-5 30 3 *" },
    { case: 'DST', tz: "Europe/Madrid", expression: "0 2 30 3 *" },
    { case: 'DST', tz: "Europe/Madrid", expression: "30 2 30 3 *" },
    { case: 'DST', tz: "Europe/Madrid", expression: "30 3 30 3 *" },
    { case: 'DST', tz: "Europe/Madrid", expression: "30 3 30 3 *" },
    { case: 'DST', tz: "Europe/Madrid", expression: "0 1-5 30 3 *" },
    // Edges
    { case: 'Edges', tz: "UTC", expression: "0 0 1 3 *" },
    { case: 'Edges', tz: "UTC", expression: "0 23 31 * *" },
    { case: 'Edges', tz: "UTC", expression: "59 23 31 * *" },
    // Inter month jobs
    { case: 'Inter month', tz: "UTC", expression: "0 0 1 * *" },
  ])(
    "should parse cron expressions correctly for case $case in $tz with $expression",
    ({ tz, expression }) => {
      const occurrences = getOccurrecesRelevantForMonth(
        expression,
        new Date("2025-03-14T04:00:00"),
        tz
      );

      expect(occurrences).toMatchSnapshot();
    }
  );
  it("should throw an error if the cron expression is invalid", () => {
    expect(() => getOccurrecesRelevantForMonth("invalid", new Date("2025-03-14T04:00:00"), "UTC")).toThrow();
  });
});

describe("validateCronExpression", () => {
  it("should return true if the cron expression is valid", () => {
    expect(validateCronExpression("0 0/1 * * *")).toBe(true);
  });
  it("should return false if the cron expression is invalid", () => {
    expect(validateCronExpression("invalid")).toBe(false);
  });
});