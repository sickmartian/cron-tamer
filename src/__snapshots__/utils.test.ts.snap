// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in Europe/Madrid with 0 1-5 30 3 * 1`] = `
[
  "2025-03-30T01:00:00.000+01:00",
  "2025-03-30T03:00:00.000+02:00",
  "2025-03-30T04:00:00.000+02:00",
  "2025-03-30T05:00:00.000+02:00",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in Europe/Madrid with 0 2 30 3 * 1`] = `
[
  "2025-03-30T03:00:00.000+02:00",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in Europe/Madrid with 30 2 30 3 * 1`] = `
[
  "2025-03-30T03:30:00.000+02:00",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in Europe/Madrid with 30 3 30 3 * 1`] = `
[
  "2025-03-30T03:30:00.000+02:00",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in Europe/Madrid with 30 3 30 3 * 2`] = `
[
  "2025-03-30T03:30:00.000+02:00",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in UTC with 0 1-5 30 3 * 1`] = `
[
  "2025-03-30T01:00:00.000Z",
  "2025-03-30T02:00:00.000Z",
  "2025-03-30T03:00:00.000Z",
  "2025-03-30T04:00:00.000Z",
  "2025-03-30T05:00:00.000Z",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in UTC with 0 2 30 3 * 1`] = `
[
  "2025-03-30T02:00:00.000Z",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in UTC with 0 3 30 3 * 1`] = `
[
  "2025-03-30T03:00:00.000Z",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in UTC with 30 2 30 3 * 1`] = `
[
  "2025-03-30T02:30:00.000Z",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case DST in UTC with 30 3 30 3 * 1`] = `
[
  "2025-03-30T03:30:00.000Z",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case Edges in UTC with 0 0 1 3 * 1`] = `
[
  "2025-03-01T00:00:00.000Z",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case Edges in UTC with 0 23 31 * * 1`] = `
[
  "2025-03-31T23:00:00.000Z",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case Edges in UTC with 59 23 31 * * 1`] = `
[
  "2025-03-31T23:59:00.000Z",
]
`;

exports[`getOccurrecesRelevantForMonth should parse cron expressions correctly for case Inter month in UTC with 0 0 1 * * 1`] = `
[
  "2025-02-01T00:00:00.000Z",
  "2025-03-01T00:00:00.000Z",
]
`;
